Adding automated unit testing to CI pipeline

MilanDonhowe opened the pull request:
This PR addresses #137. 

### 1. Quick Summary of Changes:
This PR adds jest configurations for front-end & back-end unit testing, and it also re-configures the `.travis.yml` to first run all the unit tests, and then to build/deploy the project.

The unit tests run on [Jest v.24](https://archive.jestjs.io/docs/en/24.x/getting-started.html) and have been organized into the `/tests/` folders in the frontend & backend code sections.  The jest test runner is configured to run all test suites of the form: `<filename>.spec.js`, and mock data has been stored in the `tests/assertedData/` folders.

The front-end unit tests are heavily dependent on [Vue-test-utils](https://vue-test-utils.vuejs.org/) which allows Jest to test vue stuff. 

### 2. Itemized List of Changes:
**Added Unit Tests (Frontend):**
* `datastore.spec.js`: makes sure that the persistent caching system works
* `map.module.spec.js`: makes sure the map module can load the `allbuildings` api request & dynamically load the corresponding modules.
* `building_list.spec.js`: makes sure the building list tab's selection features works.

**Added Unit Tests (Backend):**
* `building.spec.js`: makes sure that the `/allbuildings` handler returns data (& has valid CORS headers)
* `campaign.spec.js`: makes sure that the `/campaigns` handler returns valid campaign data (& has CORS headers) 
* `meter.spec.js`: makes sure that the meter data handler returns meter data (& has CORS headers)
* `now.spec.js`: makes sure that the `/systemtime` handler returns a valid timestamp (& has CORS headers)

**Added Utility Functions (Backend)**:
* `cors_test_utility.js`: this is a quick script I wrote to check the validity of a lambda function's response to make sure its CORS compliant.
* `csv_to_obj.js`: this script exports a function which formats a .csv file to the javascript array object thing which lambda functions expect to get when calling `DB.query()`

**Other Misc. Changes**:
* The vuex store now gets instantiated in the `main.js` file instead of the `index.js` file for importing convenience
* Lots of mock data has been added (.csv files for mock data from RDS) & (.json files for responses from our api)
* The Lambda Common Layer is now a dependency for the backend package.json (so we can use common layer functions in unit tests)
* `fake-indexeddb` added as dev. dependency to test the data store
* To run tests `npm run test` has been added as a functioning command to both of the `package.json` files
* Added coverage badge things to README.md using `make-coverage-badge` package.  This means badges are generated statically via `npm run make-badges` and need to be committed to the repository to update.
* `sam_express.js` now loads the ARN from the template.yaml correctly for the most recent version of node 10.

jackrwoods commented on README.md:
```diff
@@ -5,6 +5,7 @@
 # OSU Energy Dashboard
 [![BrowserStack Status](https://www.browserstack.com/automate/badge.svg?badge_key=eGdPekhGQnBGWWtHSGJRK2pORVJKVnNhUEdLbmtIcHpXRzMvaEliS2E4UT0tLVNMcy85bE9uS1dYN3lTQzAza1lkOUE9PQ==--7b21bb45dd90d0a6579467798b97153019cb4d24)](https://www.browserstack.com/automate/public-build/eGdPekhGQnBGWWtHSGJRK2pORVJKVnNhUEdLbmtIcHpXRzMvaEliS2E4UT0tLVNMcy85bE9uS1dYN3lTQzAza1lkOUE9PQ==--7b21bb45dd90d0a6579467798b9 ...
 [![Build Status](https://travis-ci.com/OSU-Sustainability-Office/energy-dashboard.svg?branch=master)](https://travis-ci.com/OSU-Sustainability-Office/energy-dashboard)
+![Frontend Test Coverage](./coverage/badge.svg) ![Backend Test Coverage](./backend/coverage/badge.svg)
```
This is a nice touch!

jackrwoods commented on backend/coverage/coverage-summary.json:
```diff
@@ -0,0 +1,13 @@
+{"total": {"lines":{"total":1403,"covered":754,"skipped":0,"pct":53.74},"statements":{"total":1403,"covered":754,"skipped":0,"pct":53.74},"functions":{"total":56,"covered":23,"skipped":0,"pct":41.07},"branches":{"total":65,"covered":46,"skipped":0,"pct":70.77}}
```
Can we add this file to the .gitignore, or is this necessary for the coverage badge?
When I think of code coverage I see two primary use cases:

1. As a developer, I want to view the coverage of a new test I added.
2. As a developer, I want to be notified if changes reduce the code coverage beyond a given threshold. For example, a build should fail if the coverage is less than x%.

jackrwoods commented on backend/jest.config.js:
```diff
@@ -0,0 +1,197 @@
+/*
+ * For a detailed explanation regarding each configuration property, visit:
+ * https://jestjs.io/docs/en/configuration.html
```
👍

jackrwoods commented on backend/jest.config.js:
```diff
@@ -0,0 +1,197 @@
+  coverageReporters: [
+    'json-summary'
+  ],
+  //   "json",
```
Do we need to leave this code commented out here?

jackrwoods commented on backend/jest.config.js:
```diff
@@ -0,0 +1,197 @@
+  //   "clover"
+  // ],
+
+  // An object that configures minimum threshold enforcement for coverage results
```
If these configuration options are documented at the URL listed at the top of this document, I don't think we need to leave it all commented out here. As a developer in the future, I'll probably need to reference the documentation anyways when modifying this config file.

MilanDonhowe replied:
Yeah, jest auto-generates the config file with those comments and I just neglected to remove them; I'll clean it up.

jackrwoods commented on backend/tests/building.spec.js:
```diff
@@ -0,0 +1,84 @@
+
+const AllBuildings = require('../app/building.js')
+
+describe('Testing allbuildings endpoint...', () => {
```
Is there a benefit to having a separate file for each endpoint? I feel that we can eliminate code duplication by running tests against allBuildings, campaigns, etc in one file (and use a smarter mock that grabs the correct data based on the request url).

jackrwoods replied:
Perhaps we can group these somehow- maybe by the VueX module they correspond to? What are your thoughts @MilanDonhowe

MilanDonhowe replied:
Yeah, there's not a huge benefit in keeping them separated beyond just making the database mocks easier to write.  We could try grouping them based on their corresponding VueX module, my only concern would be that in the event that we modified or drastically re-structured the VueX store (which probably won't happen, but is worth mentioning) it would be a bit of a hassle to re-structure the backend tests.  For now, I'll go ahead and group by VueX modules and see how it works.

I'll also eliminate a lot of duplicate code by adding a [setup script ](https://jestjs.io/docs/configuration#setupfiles-array) to `require()` a lot of redundant modules.

jackrwoods commented on src/main.js:
```diff
@@ -2,29 +2,27 @@
  * @Date:   2019-01-04T10:08:23-08:00
  * @Email:  brogan.miner@oregonstate.edu
- * @Last modified by:   Brogan
+ * @Last modified by:   Milan
```
I believe there is an Atom (and possibly vscode) extension that modifies these headers automatically, with fresh dates and your name. Personally, I think we can slowly just delete these tags because git tracks this information on a commit-by-commit basis.

MilanDonhowe commented:
Alright, here's the changes I've made to the PR:

- Removed superfluous comments from jest.config files
- Added smarter Database mock via a sqlite DB
- Reorganized backend unit tests into module grouping
- Reduced unit test code by using a Jest setup script

jackrwoods commented on backend/tests/buildTestDatabase.js:
```diff
@@ -0,0 +1,128 @@
+        console.log('built test database!')
+    })
+})
+DB.close();
```
I think using a SQLite DB is a fantastic idea!
